File: java/152.MaximumProductSubarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int midRes=nums[0];
        int midNegRes=nums[0];
        int res=nums[0];
        int midMin;
        int midMax;
        for(int i=1;i<nums.length;i++){ 
            midMin = Math.min(midNegRes*nums[i],midRes*nums[i]);
            midMax = Math.max(midNegRes*nums[i],midRes*nums[i]);
            midNegRes=Math.min(midMin,nums[i]);
            midRes=Math.max(midMax,nums[i]);
            res=Math.max(res,midRes);
        }
        return res;
    }
}

File: Leetcode/cpp/152. Maximum Product Subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int res = INT_MIN;
        int curSum1 = 1;
        int curSum2 = 1;
        for (int num : nums) {
            int mid=curSum2;
            curSum2= min(min(curSum2 * num,curSum1 * num),num);
            curSum1 = max(max(mid * num,curSum1 * num),num);
            if(curSum2>curSum1)swap(curSum2,curSum1);
            res = max(res, curSum1);
        }
        return res;
    }
};
